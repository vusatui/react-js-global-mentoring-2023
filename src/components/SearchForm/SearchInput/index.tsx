import {ChangeEvent, CSSProperties} from "react";

import {StyledInput} from "./styled";

interface Props {
    className?: string,
    placeholder?: string
    style?: CSSProperties
    value: string
    onChange: (v: ChangeEvent<HTMLInputElement>) => void
}

const SearchInput = ({
    placeholder = "What do you want to watch?",
    style = undefined,
    className = undefined,
    value,
    onChange,
}: Props) => (
    <StyledInput
        className={className}
        style={style}
        value={value}
        placeholder={placeholder}
        onChange={onChange}
    />
);

export default SearchInput;