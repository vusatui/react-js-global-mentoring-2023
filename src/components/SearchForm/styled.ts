"use client";

import styled from "styled-components";

import SearchInput from "./SearchInput";

export const StyledForm = styled.form`
  display: flex;
  flex-flow: row nowrap;
`;

export const StyledSearchInput = styled(SearchInput)`
  margin-right: 14px;
  width: 713px;
`;


