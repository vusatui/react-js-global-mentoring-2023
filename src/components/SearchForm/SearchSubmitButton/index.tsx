import {StyledButton} from "./styled";

interface SearchSubmitButtonProps {
    className?: string,
    text?: string
}

const SearchSubmitButton = ({
    className,
    text = "SEARCH",
}: SearchSubmitButtonProps) => (
    <StyledButton
        type="submit"
        className={className}
    >
        {text}
    </StyledButton>
);

export default SearchSubmitButton;