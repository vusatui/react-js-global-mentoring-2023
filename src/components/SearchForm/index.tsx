"use client";

import {ChangeEvent, FormEvent, useState} from "react";

import {StyledForm, StyledSearchInput} from "./styled";

import SearchSubmitButton from "./SearchSubmitButton";

interface SearchFormProps {
    placeholder?: string
    initialValue?: string
    onSearch: (v: string) => void
}

const SearchForm = ({
    placeholder,
    initialValue = "",
    onSearch,
}: SearchFormProps) => {
    const [searchValue, setSearchValue] = useState(initialValue)

    const handleChange = (e: ChangeEvent<HTMLInputElement>) => setSearchValue(e.target.value);

    const handleFormSubmit = (e: FormEvent) => {
        e.preventDefault();
        onSearch(searchValue);
    };

    return (
        <StyledForm onSubmit={handleFormSubmit}>
            <StyledSearchInput
                value={searchValue}
                placeholder={placeholder}
                onChange={handleChange}
            />
            <SearchSubmitButton />
        </StyledForm>
    );
};

export default SearchForm;