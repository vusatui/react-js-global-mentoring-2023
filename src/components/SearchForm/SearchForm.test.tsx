import SearchForm from "./index";
import {act, render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const INITIAL_VALUE: string = "James Bond";
const UPDATED_VALUE: string = "Star Wars";

const mockHandleSearch = jest.fn();

const Component = () => <SearchForm initialValue={INITIAL_VALUE} onSearch={mockHandleSearch} />;

describe("SearchForm", () => {

    it("Should renders correctly", () => {
        const { asFragment } = render(<Component />);
        expect(screen.getByDisplayValue(INITIAL_VALUE)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("Test 'onSelect' passes correct genre in arguments", () => {
        render(<Component />);
        const input = screen.getByTestId("search-input");
        const button = screen.getByTestId("search-button");
        act(() => {
            userEvent.clear(input);
            userEvent.type(input, UPDATED_VALUE);
            userEvent.click(button);
        });
        expect(mockHandleSearch).toBeCalledWith(UPDATED_VALUE);
    });

    it("Test that after typing to the input and pressing Enter key, the 'onChange' prop is called with proper value", () => {
        render(<Component />);
        const input = screen.getByTestId("search-input");
        act(() => {
            userEvent.clear(input);
            userEvent.type(input, UPDATED_VALUE)
            userEvent.keyboard("{enter}");
        });
        expect(mockHandleSearch).toBeCalledWith(UPDATED_VALUE);
    });

});