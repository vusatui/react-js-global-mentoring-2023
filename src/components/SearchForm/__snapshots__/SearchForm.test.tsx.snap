// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Should renders correctly 1`] = `
<DocumentFragment>
  <form
    class="sc-gueYoa hVeTQV"
  >
    <input
      class="sc-beqWaB iYpUnI sc-dmqHEX dNlPqL"
      data-testid="search-input"
      placeholder="What do you want to watch?"
      value="James Bond"
    />
    <button
      class="sc-hLseeU cvQoS"
      data-testid="search-button"
      type="submit"
    >
      SEARCH
    </button>
  </form>
</DocumentFragment>
`;
